backbone.js

```diff
@@ -1511,13 +1535,18 @@
 
     // Convert a route string into a regular expression, suitable for matching
     // against the current location hash.
-    _routeToRegExp: function(route) {
+    _routeToRegExp: function(route, paramNames) {
       route = route.replace(escapeRegExp, '\\$&')
                    .replace(optionalParam, '(?:$1)?')
                    .replace(namedParam, function(match, optional) {
-                     return optional ? match : '([^/?]+)';
+                     if (optional) return match;
```

Comment:
Yep, they are being recorded. At this point, this replace handles all named parameters, including optional ones, as previous replace already appends : to the name:

'named/optional/(y:z)'
  .replace(escapeRegExp, '\\$&')                    // 'named/optional/(y:z)''
  .replace(optionalParam, '(?:$1)?')                // 'named/optional/(?:y:z)?'
  .replace(namedParam, function(match, optional) {
    // this function gets called twice
    // 1st
    //   match: '(?:y'
    //   optional: '(?'
    // 2nd
    //   match: ':z'
    //   optional: undefined
  });